openpaas-james/mailbox/encrypted/api/src/main/scala/com/linagora/openpaas/encrypted/KeystoreManagerContract.scala

```diff
@@ -0,0 +1,168 @@
+
+    val keyId1: KeyId = SMono.fromPublisher(keyStoreManager.save(BOB, payload1)).block()
+    SMono.fromPublisher(keyStoreManager.save(BOB, payload2)).block()
+    val key1: PublicKey = PublicKey(keyId1, payload1)
+
+    assertThat(SMono.fromPublisher(keyStoreManager.retrieveKey(BOB, keyId1)).block())
+      .isEqualTo(key1)
+  }
+
+  @Test
+  def listPublicKeysShouldReturnEmptyWhenUserDoesNotExist(): Unit = {
+    assertThat(SFlux.fromPublisher(keyStoreManager.listPublicKeys(BOB)).collectSeq().block().asJava)
+      .isEmpty()
+  }
+
+  @Test
+  def deleteAllShouldSucceed(): Unit = {
+    val payload1: Array[Byte] = ClassLoader.getSystemClassLoader.getResourceAsStream("gpg.pub").readAllBytes()
+    val payload2: Array[Byte] = ClassLoader.getSystemClassLoader.getResourceAsStream("gpg2.pub").readAllBytes()
+
+    SMono.fromPublisher(keyStoreManager.save(BOB, payload1)).block()
+    SMono.fromPublisher(keyStoreManager.save(BOB, payload2)).block()
+
+    SMono.fromPublisher(keyStoreManager.deleteAll(BOB)).block()
+
+    assertThat(SFlux.fromPublisher(keyStoreManager.listPublicKeys(BOB)).collectSeq().block().asJava)
+      .isEmpty()
+  }
+
+  @Test
+  def deleteAllShouldLeaveOtherUserKeysIntact(): Unit = {
+    val payload1: Array[Byte] = ClassLoader.getSystemClassLoader.getResourceAsStream("gpg.pub").readAllBytes()
+    val payload2: Array[Byte] = ClassLoader.getSystemClassLoader.getResourceAsStream("gpg2.pub").readAllBytes()
+
+    SMono.fromPublisher(keyStoreManager.save(BOB, payload1)).block()
+    val keyId: KeyId = SMono.fromPublisher(keyStoreManager.save(ALICE, payload2)).block()
+    val key: PublicKey = PublicKey(keyId, payload2)
+
+    SMono.fromPublisher(keyStoreManager.deleteAll(BOB)).block()
+
+    assertThat(SFlux.fromPublisher(keyStoreManager.listPublicKeys(ALICE)).collectSeq().block().asJava)
+      .contains(key)
+  }
+
+  @Test
+  def deleteAllShouldSucceedWhenUserHasNoKey(): Unit = {
+    assertThatCode(() => SMono.fromPublisher(keyStoreManager.deleteAll(BOB)).block())
+      .doesNotThrowAnyException()
+  }
+
+  @Test
+  def deleteShouldSucceed(): Unit = {
+    val payload1: Array[Byte] = ClassLoader.getSystemClassLoader.getResourceAsStream("gpg.pub").readAllBytes()
+
+    val keyId: KeyId = SMono.fromPublisher(keyStoreManager.save(BOB, payload1)).block()
+
+    SMono.fromPublisher(keyStoreManager.delete(BOB, keyId)).block()
+
+    assertThat(SFlux.fromPublisher(keyStoreManager.listPublicKeys(BOB)).collectSeq().block().asJava)
+      .isEmpty()
+  }
+
+  @Test
+  def deleteShouldSucceedWhenUserHasNoKey(): Unit = {
+    val payload1: Array[Byte] = ClassLoader.getSystemClassLoader.getResourceAsStream("gpg.pub").readAllBytes()
+    val keyId: KeyId = KeyId.fromPayload(payload1)
+
+    assertThatCode(() => SMono.fromPublisher(keyStoreManager.delete(BOB, keyId)).block())
+      .doesNotThrowAnyException()
+  }
+
+  @Test
+  def deleteShouldThrowWhenKeyIdNotFound(): Unit = {
```

Comment:
This is not idempotent.

I would be in favor of a `deleteShouldBeIdempotent` which means "should not throw"